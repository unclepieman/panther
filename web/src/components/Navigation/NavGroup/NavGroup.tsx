/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import { Box, AbstractButton, Icon, IconProps, Collapse } from 'pouncejs';
import React from 'react';

type NavGroupProps = {
  icon: IconProps['type'];
  label: string;
  active?: boolean;
  onSelect: () => void;
  children?: React.ReactNode;
};

const NavGroup: React.FC<NavGroupProps> = ({ icon, label, active, onSelect, children }) => {
  return (
    <Box
      fontSize="medium"
      borderRadius="small"
      backgroundColor={active ? 'navyblue-500' : 'transparent'}
    >
      <AbstractButton
        px={4}
        py={3}
        display="flex"
        onClick={onSelect}
        alignItems="center"
        width="100%"
        _hover={{
          color: 'gray-50',
          backgroundColor: 'navyblue-500',
        }}
      >
        <Icon type={icon} size="medium" mr={3} />
        <Box>{label}</Box>
        <Icon
          transition="transform 200ms cubic-bezier(0.0, 0, 0.2, 1) 80ms"
          transform={active ? 'rotate(180deg)' : 'rotate(0deg)'}
          type="chevron-down"
          size="medium"
          justifySelf="flex-end"
          ml="auto"
        />
      </AbstractButton>
      <Collapse open={active}>
        <Box px={2} pb={2}>
          {children}
        </Box>
      </Collapse>
    </Box>
  );
};

export default NavGroup;
