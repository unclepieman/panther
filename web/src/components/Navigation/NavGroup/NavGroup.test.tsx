/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import React from 'react';
import { render, fireEvent, waitFor } from 'test-utils';
import { Box } from 'pouncejs';
import NavGroup from './index';

const Nav = ({ onSelect }) => {
  const [active, setActive] = React.useState(false);
  const onSelectCallback = () => {
    onSelect();
    setActive(!active);
  };
  return (
    <NavGroup active={active} onSelect={onSelectCallback} label="Test NavGroup" icon="alert">
      <Box>Child 1</Box>
      <Box>Child 2</Box>
    </NavGroup>
  );
};

describe('NavGroup', () => {
  it('could render a collapsing group of items ', async () => {
    const onSelect = jest.fn();
    const { container, getByText, queryByText } = render(<Nav onSelect={onSelect} />);

    expect(container).toMatchSnapshot();
    expect(queryByText('Child 1')).not.toBeInTheDocument();
    expect(queryByText('Child 2')).not.toBeInTheDocument();

    fireEvent.click(getByText('Test NavGroup'));

    await waitFor(() => {
      expect(getByText('Child 1')).toBeInTheDocument();
    });
    expect(getByText('Child 2')).toBeInTheDocument();
    expect(onSelect).toHaveBeenCalled();
  });
});
