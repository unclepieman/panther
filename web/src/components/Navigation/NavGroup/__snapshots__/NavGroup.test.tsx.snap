// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavGroup could render a collapsing group of items  1`] = `
.panther-4 {
  font-size: 0.875rem;
  border-radius: 2px;
  background-color: transparent;
}

.panther-3 {
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 12px;
  padding-bottom: 12px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  cursor: pointer;
  color: #f6f6f6;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: transparent;
  -webkit-transition: all 0.1s linear;
  transition: all 0.1s linear;
}

.panther-3:hover {
  color: #f6f6f6;
  background-color: #1d2a3f;
}

.panther-0 {
  display: inline-block;
  vertical-align: sub;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 20px;
  height: 20px;
  color: currentColor;
  margin-right: 12px;
}



.panther-2 {
  display: inline-block;
  vertical-align: sub;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 20px;
  height: 20px;
  color: currentColor;
  -webkit-transition: -webkit-transform 200ms cubic-bezier(0.0, 0, 0.2, 1) 80ms;
  transition: transform 200ms cubic-bezier(0.0, 0, 0.2, 1) 80ms;
  -webkit-transform: rotate(0deg);
  -moz-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
  justify-self: flex-end;
  margin-left: auto;
}

<div>
  <header
    id="main-header"
  />
  <div
    class="panther-4"
  >
    <button
      class="panther-3"
      type="button"
    >
      <svg
        class="panther-0"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M4,19V5H15.5V9h2V5a2,2,0,0,0-2-2H4A2,2,0,0,0,2,5V19a2,2,0,0,0,2,2H9V19H4Z"
        />
        <path
          d="M20 11H13a2 2 0 00-2 2v7a2 2 0 002 2h7a2 2 0 002-2V13A2 2 0 0020 11zm-2.75 9h-1.5V18.5h1.5zm0-3.15h-1.5v-4h1.5zM6 15.01H9V17.009999999999998H6zM6 11H9V13H6zM6 7H13.5V9H6z"
        />
      </svg>
      <div
        class="panther-1"
      >
        Test NavGroup
      </div>
      <svg
        class="panther-2"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M7.41 8.58997L12 13.17L16.59 8.58997L18 9.99997L12 16L6 9.99997L7.41 8.58997Z"
        />
      </svg>
    </button>
  </div>
  <footer
    id="footer"
  />
</div>
`;
